### PR TITLE
Fix eager loading

### PR DESCRIPTION
**Pull request by smn**

### CODE REVIEW

File: test/expression/eval_test.exs [hunk 1]
```diff
@@ -27,41 +27,49 @@ defmodule Expression.EvalTest do
...
+  end
+
+  test "if" do
+    {:ok, ast, "", _, _, _} =
+      Parser.parse("@if(image_response.status == 200,\nimage_response.body.id,\nfalse)")
+
+    assert false ==
+             Eval.eval!(ast, %{
+               "image_response" => %{"status" => 500, "body" => "Internal Server Error"}
+             })
```
Comment (smn) [inline test/expression/eval_test.exs:40]: @santiagocardo all this work just to make sure this doesn't crash ...

**Comment by smn** [comment]
https://github.com/turnhub/flow_runner/pull/47 and https://github.com/turnhub/stacks_dsl/pull/23 use these changes, once released & publish I'll update the mix.exs in those PRs.

File: lib/expression/callbacks.ex [hunk 2]
```diff
@@ -991,134 +928,147 @@ defmodule Expression.Callbacks do
...
-  def word_slice(_ctx, binary, start, stop, by_spaces) when stop < 0 do
+  def word_slice(ctx, binary, start, stop, by_spaces) do
+    [binary, start, stop, by_spaces] = eval_args!([binary, start, stop, by_spaces], ctx)
     splitter = if(by_spaces, do: " ", else: @punctuation_pattern)
 
-    binary
-    |> String.split(splitter)
-    |> Enum.slice((start - 1)..(stop - 1))
-    |> Enum.join(" ")
+    cond do
```
Comment (santiagocardo) [inline lib/expression/callbacks.ex]: not that is expensive in this situation, but we can use `case` instead so we keep using conventions in favor of more performant code.

Here and the other places where this is being used

File: lib/expression/callbacks.ex [hunk 3]
```diff
@@ -991,134 +928,147 @@ defmodule Expression.Callbacks do
...
-      iex> Expression.Callbacks.isnumber(%{}, "a")
+      iex> Expression.evaluate!("@isnumber(\\"a\\")")
       false
 
   """
-  def isnumber(_ctx, var) when is_float(var) or is_integer(var), do: true
+  def isnumber(ctx, var) do
+    var = eval!(var, ctx)
+
+    cond do
```
Comment (santiagocardo) [inline lib/expression/callbacks.ex]: same here

File: lib/expression/callbacks.ex [hunk 4]
```diff
@@ -991,134 +928,147 @@ defmodule Expression.Callbacks do
...
       false
-      iex> Expression.Callbacks.isstring(%{}, 1)
+      iex> Expression.evaluate!("@isstring(1)")
       false
-      iex> Expression.Callbacks.isstring(%{}, Decimal.new("1.0"))
+      iex> Expression.evaluate!("@isstring(d)", %{"d" => Decimal.new("1.0")})
       false
   """
-  def isstring(_ctx, binary), do: is_binary(binary)
+  def isstring(ctx, binary), do: is_binary(eval!(binary, ctx))
```
Comment (santiagocardo) [inline lib/expression/callbacks.ex:1081]: can we rename this to `string?` or `is_string`

File: lib/expression/callbacks.ex [hunk 5]
```diff
@@ -21,6 +21,14 @@ defmodule Expression.Callbacks do
   underscore.
   """
 
+  def eval!(ast, ctx) do
+    ast
+    |> Expression.Eval.eval!(ctx, __MODULE__)
+    |> Expression.Eval.not_founds_as_nil()
+  end
+
+  def eval_args!(args, ctx), do: Enum.map(args, &eval!(&1, ctx))
```
Comment (santiagocardo) [inline lib/expression/callbacks.ex]: could you please add some docs for these functions?

**Review by santiagocardo — APPROVED** [review]
LGTM. huge job, @smn!

File: lib/expression/callbacks.ex (hunk 4 above)
Comment (santiagocardo) [reply lib/expression/callbacks.ex:1081]: omit it if this is just following the `Expressions` specs

Comment (smn) [reply lib/expression/callbacks.ex:1081]: the FLOIP spec standard dictates these function names ... 🤷

File: lib/expression/callbacks.ex (hunk 2 above)
Comment (smn) [reply lib/expression/callbacks.ex]: done!

File: lib/expression/callbacks.ex (hunk 3 above)
Comment (smn) [reply lib/expression/callbacks.ex]: done!

File: lib/expression/callbacks.ex (hunk 5 above)
Comment (smn) [reply lib/expression/callbacks.ex]: done!

**Pull request merged**